### PR TITLE
VPN-6285 - Add metric to record data transmitted per connection session

### PR DESCRIPTION
**Pull request by brizental**

### CODE REVIEW

**Comment by brizental** [comment]
!! Reminder: it is your responsibility to complete and check the correctness of
!! this automatically-generated request skeleton before requesting Data
!! Collection Review. See https://wiki.mozilla.org/Data_Collection for details.

DATA REVIEW REQUEST
1. What questions will you answer with this data?

How much bandwidth goes thorugh the Mozilla VPN tunnel.

2. Why does Mozilla need to answer these questions? Are there benefits for users?
   Do we need this information to address product or business requirements?

We want to understand the bandwidth usage on the Mozilla VPN.

This will help inform numerous technical discussion happening on the team.

3. What alternative methods did you consider to answer these questions?
   Why were they not sufficient?

None.

4. Can current instrumentation answer these questions?

No.

5. List all proposed measurements and indicate the category of data collection for each
   measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement Name | Measurement Description | Data Collection Category | Tracking Bug
---------------- | ----------------------- | ------------------------ | ------------
`connection.bandwidth_usage` | The amount of bandwidth sent / received through the VPN tunnel.  The is accumulated periodically (TBD) while the VPN is turned on.  | technical | https://github.com/mozilla-mobile/mozilla-vpn-client/issues/VPN-5759

6. Please provide a link to the documentation for this data collection which
   describes the ultimate data set in a public, complete, and accurate way.

This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).

7. How long will this data be collected?

This collection will be collected permanently.
vpn-telemetry@mozilla.com, brizental@mozilla.com will be responsible for the permanent collections.

8. What populations will you measure?

All channels, countries, and locales. No filters.

9. If this data collection is default on, what is the opt-out mechanism for users?

These collections are Glean. The opt-out can be found in the product's preferences.

10. Please provide a general description of how you will analyze this data.

Looker / Redash.

11. Where do you intend to share the results of your analysis?

Within the Mozilla VPN team.

12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
    are proposing to use for this data collection?

No.

cc @travis79

**Review by travis79 — APPROVED** [review]
> !! Reminder: it is your responsibility to complete and check the correctness of !! this automatically-generated request skeleton before requesting Data !! Collection Review. See https://wiki.mozilla.org/Data_Collection for details.
> 
> DATA REVIEW REQUEST
> 
>     1. What questions will you answer with this data?
> 
> 
> How much bandwidth goes thorugh the Mozilla VPN tunnel.
> 
>     2. Why does Mozilla need to answer these questions? Are there benefits for users?
>        Do we need this information to address product or business requirements?
> 
> 
> We want to understand the bandwidth usage on the Mozilla VPN.
> 
> This will help inform numerous technical discussion happening on the team.
> 
>     3. What alternative methods did you consider to answer these questions?
>        Why were they not sufficient?
> 
> 
> None.
> 
>     4. Can current instrumentation answer these questions?
> 
> 
> No.
> 
>     5. List all proposed measurements and indicate the category of data collection for each
>        measurement, using the Firefox data collection categories found on the Mozilla wiki.
> 
> 
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug
> `connection.bandwidth_usage` 	The amount of bandwidth sent / received through the VPN tunnel. The is accumulated periodically (TBD) while the VPN is turned on. 	technical 	https://github.com/mozilla-mobile/mozilla-vpn-client/issues/VPN-5759
> 
>     6. Please provide a link to the documentation for this data collection which
>        describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).
> 
>     7. How long will this data be collected?
> 
> 
> This collection will be collected permanently. [vpn-telemetry@mozilla.com](mailto:vpn-telemetry@mozilla.com), [brizental@mozilla.com](mailto:brizental@mozilla.com) will be responsible for the permanent collections.
> 
>     8. What populations will you measure?
> 
> 
> All channels, countries, and locales. No filters.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> These collections are Glean. The opt-out can be found in the product's preferences.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Looker / Redash.
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Within the Mozilla VPN team.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you
>         are proposing to use for this data collection?
> 
> 
> No.
> 
> cc @travis79

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/).

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, through the data preferences in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

Permanent collection to be monitored by brizental

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 1, Technical data

5) Is the data collection request for default-on or default-off?

Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice?

Yes

8) Does the data collection use a third-party collection tool?

No

## Result

data-review+

File: src/controller.cpp [hunk 1]
```diff
@@ -837,6 +838,13 @@ void Controller::statusUpdated(const QString& serverIpv4Gateway,
                                const QString& deviceIpv4Address,
                                uint64_t txBytes, uint64_t rxBytes) {
   logger.debug() << "Status updated";
+
```
Comment (oskirby) [inline src/controller.cpp:841]: My understanding of the `Controller` and how it interracts with the rest of the VPN is that we only call the `getStatus()` method once after connection to fetch the gateway addresses and kick off the connection health monitoring. This means that we aren't really going to accumulate meaningful data as the counter will always be somewhere around zero.

What we need is something to fetch this data periodically, or to gather the data upon connection deactivation.

**Review by oskirby — COMMENTED** [review]
I think the metrics look good, but I am concerned that as written this does not generate correct data.

File: src/controller.cpp (hunk 1 above)
Comment (brizental) [reply src/controller.cpp:841]: Yeap, I thought I was being smart here, but I could not have been more wrong. You are right and I will change this.

**Pull request closed without merging**